pydantic/main.py

```diff
@@ -421,14 +423,21 @@ def __get_validators__(cls) -> 'CallableGenerator':
         yield cls.validate
 
     @classmethod
-    def validate(cls: Type['Model'], value: Union['DictStrAny', 'Model']) -> 'Model':
+    def validate(cls: Type['Model'], value: Any) -> 'Model':
         if isinstance(value, dict):
             return cls(**value)
         elif isinstance(value, cls):
             return value.copy()
-        else:
+        elif hasattr(value, '__iter__'):
             with change_exception(DictError, TypeError, ValueError):
-                return cls(**dict(value))  # type: ignore
+                return cls(**dict(value))
+        else:
+            with change_exception(DictError, TypeError, ValueError, AttributeError):
+                return cls(**cls._decompose_class(value))
+
+    @classmethod
+    def _decompose_class(cls, obj: Any) -> 'DictStrAny':
+        return obj.__dict__
```

Comment:
If you don't want to (or can't) use `.__dict__`, having this as a separate function means you can override `_decompose_class` and build a dict some other way.